File: top_interview_questions/java/46_permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List ans = new ArrayList<ArrayList<Integer>> ();
        boolean[] visited = new boolean[nums.length];
        backtrack(ans, nums, new LinkedList(), visited);
        return ans;
    }

    public void backtrack(List<List<Integer>> ans, int[] nums, LinkedList<Integer> temp,  boolean[] visited) {
        if (temp.size() == nums.length) {
            ans.add(new ArrayList(temp));
        }
        for (int i = 0; i < nums.length; i++) {
            if (!visited[i]) {
                temp.addLast(nums[i]);
                visited[i] = true;
                backtrack(ans, nums, temp, visited);
                temp.removeLast();
                visited[i] = false;
            }
        }
    }
}

File: top_interview_questions/javascript/46_permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const ans = new Array();
    const visited = new Array(nums.length).fill(0);
    backtrack(ans, nums, [], visited);
    return ans;

};

function backtrack(ans, nums, temp, visited) {
    if (temp.length === nums.length) {
        ans.push([...temp]);
        return;
    }
    for (let i = 0; i < nums.length; i++) {
        if (!visited[i]) {
            temp.push(nums[i]);
            visited[i] = 1;
            backtrack(ans, nums, temp, visited);
            temp.pop();
            visited[i] = 0;
        }
    }
}